File: cs/MatchableString/MatchableString/SuffixArray/MatchAbleStringSA.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MatchableString.SuffixArray
{
    public class MatchableStringSA : MatchableString
    {
        private int[] substringArray;
        private bool isPreProcessed;

        private int binarySearch(int min, int max, ComparerBase comparer)
        {
            int num = (min + max) / 2;

            switch (comparer.Compare(num))
            {
                case 0:
                    return num;
                case -1:
                    return binarySearch(num + 1, max, comparer);
                case 1:
                    return binarySearch(min, num, comparer);
                default:
                    throw new NotImplementedException("Not possible to get here - theoretically.");
            }
        }

        public MatchableStringSA(string orig) : base(orig)
        {
            substringArray = new int[Str.Length];

            for (int i = 0; i < substringArray.Length; i++)
            {
                substringArray[i] = i;
            }

            isPreProcessed = false;
        }

        public override void PreProcess(bool parallel)
        {
            IComparer<int> comp = new MergeSort.Comparer(Str);

            if (!parallel)
            {
                Array.Sort(substringArray, comp);
            }
            else
            {
                MergeSort.ParallelSort(substringArray, comp);
            }

            isPreProcessed = true;
        }

        public override List<int> Match(Pattern pattern, bool parallel)
        {
            if (!isPreProcessed)
            {
                throw new InvalidOperationException("This MatchAbleString hasn't been preprocessed yet!");
            }

            ComparerBase foc = new FirstOccurenceComparer(Str, pattern.Str, substringArray);
            ComparerBase loc = new LastOccurenceComparer(Str, pattern.Str, substringArray);

            int first = binarySearch(0, Str.Length, foc);
            int last = binarySearch(0, Str.Length, loc);

            if (first == -1)
            {
                return new List<int>();
            }
            else
            {
                List<int> tmp = new List<int>();

                for (int i = first; i <= last; i++)
                {
                    tmp.Add(substringArray[i]);
                }

                tmp.Sort();

                return tmp;
            }
        }
    }
}

File: java/MatchableString/src/matchablestring/suffixarray/MatchableStringSA.java
package matchablestring.suffixarray;

import matchablestring.MatchableString;
import matchablestring.Pattern;

import java.util.*;

public class MatchableStringSA extends MatchableString {
    private Integer[] substringArray;
    private boolean isPreProcessed;

    private int binarySearch(int min, int max, ComparerBase comparer) {
        int num = (min + max) / 2;

        switch (comparer.Compare(num)) {
            case 0:
                return num;
            case -1:
                return binarySearch(num + 1, max, comparer);
            case 1:
                return binarySearch(min, num, comparer);
            default:
                throw new UnsupportedOperationException("Not possible to get here - theoretically.");
        }
    }

    public MatchableStringSA(String s) {
        super(s);

        substringArray = new Integer[Str.length()];

        for (int i = 0; i < substringArray.length; i++) {
            substringArray[i] = i;
        }

        isPreProcessed = false;
    }

    public void preprocess(boolean parallel) {
        Comparator<Integer> comp = new MergeSort.Comparer(Str);

        if (!parallel) {
            Arrays.sort(substringArray, comp);
        } else {
            MergeSort.parallelSort(substringArray, comp);
        }

        isPreProcessed = true;
    }

    public List<Integer> match(Pattern pattern, boolean parallel) {
        if (!isPreProcessed) {
            throw new IllegalStateException("This MatchAbleString hasn't been preprocessed yet!");
        }

        ComparerBase foc = new FirstOccurenceComparer(Str, pattern.Str, substringArray);
        ComparerBase loc = new LastOccurenceComparer(Str, pattern.Str, substringArray);

        int first = binarySearch(0, Str.length(), foc);
        int last = binarySearch(0, Str.length(), loc);

        if (first == -1) {
            return new ArrayList<Integer>();
        } else {
            List<Integer> tmp = new ArrayList<Integer>();

            for (int i = first; i <= last; i++) {
                tmp.add(substringArray[i]);
            }

            Collections.sort(tmp);

            return tmp;
        }
    }
}
